Issue #14 - Add Kubernetes output option to Toolform (Minikube flavour)

NoxHarmonium opened the pull request:
### Summary

To keep things simple, all the Kubernetes deployment and service specs are concatenated into a single file. I don't believe this will cause incompatibilities with any tools (it works fine with Minikube) but it shouldn't be too hard to split it out at a later point if we want.

I have tested the output by running the seti project in Minikube.
I had to manually fix the component image names because we have changed that since the images were created but apart from that it worked well.

### Naming Changes

- Kubernetes can't automatically inspect a container to see what ports it has exposed like Docker Compose can. Therefore a new construct needed to be added to specify what ports a service exposes. We already had a `exposedPorts` property which specified ports that were exposed externally to the cluster but it was decided that the naming was confusing. Therefore the property previously named `exposedPorts` was renamed to `externalPorts`. A new property was added to specify what ports a container exposes to the internal network and it was given the name `exposedPorts` as this is the terminology used by Docker.
@timechanter mentioned that it might be worth looking into using links to provide this functionality but we decided to leave that as future work.

- Renamed the `Service` trait to `ToolFormService` to distinguish it from a Kubernetes "service". 

References issue #14

Commit:
[14] Add ability to generate Kubernetes output

- Currently only works with Minikube

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,72 @@
+    *
+    * @return the string representation of this object.
+    */
+  override def toString: String = {
```
Rename - leave `toString` as is.

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,72 @@
+  * @param containerPort  the port that is exposed by the container. This is optional and defaults to the same value as port.
+  * @param protocol       the protocol used by the port. This defaults to Tcp.
+  */
+case class PortMapping(port: Int, containerPort: Option[Int], protocol: PortProtocolType = PortProtocolType.Tcp) {
```
`final`.

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/ToolFormService.scala:
```diff
@@ -24,5 +37,5 @@ trait Service {
     *
     * @return A list of exposed ports.
```
"external ports"?

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,72 @@
+  // TODO: Is there a nicer way to deal with this kind of matching in Scala?
+  def parsePortMappingFromConfigString(value: String): PortMapping =
+    value match {
+      case parseRegex(portString, null, null) =>
```
Something like:
```
case parseRegex(portString, maybeContainerPortString, maybeProtocolString) =>
  val port = portString.toInt
  val maybeContainerPort = Option(maybeContainerPortString).map(_.toInt)
  val maybeProtocol = Option(maybeProtocolString).map(PortProtocolType.withNameInsensitive)
  PortMapping(port, maybeContainerPort, maybeProtocol)
```

NoxHarmonium replied:
Thanks! That looks much nicer.
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,72 @@
+
+object PortMapping {
+
+  private val parseRegex = """^(\d+)(?::(\d+))?(?:\/(udp|tcp))?$""".r
```
`PortMappingRegex` see https://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,72 @@
+    value match {
+      case parseRegex(portString, null, null) =>
+        val port = portString.toInt
+        new PortMapping(port, None)
```
`new` is not needed here.

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/DockerFormatting.scala:
```diff
@@ -1,32 +1,19 @@
-  def componentImageName(projectId: String, component: Component): String =
-    normaliseImageName(s"$projectId/${component.id}")
+  def formatPort(port: String): String = {
+    val normalisedPort = PortMapping.parsePortMappingFromConfigString(port).toString
```
This should be done when the project is read.

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,72 @@
+        val containerPort = containerPortString.toInt
+        val protocol      = PortProtocolType.withNameInsensitive(protocolString)
+        new PortMapping(port, Some(containerPort), protocol)
+      case _ => throw new IllegalArgumentException(s"Port value [$value] should be in the format <port>[:containerPort][/udp|/tcp]")
```
Change return type of this method to `Either`. See comment below about parsing when project configuration is read. This error should be reported during inspect, etc.

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,72 @@
+  def parsePortMappingFromConfigString(value: String): PortMapping =
+    value match {
+      case parseRegex(portString, null, null) =>
+        val port = portString.toInt
```
`toInt` is not safe. You're able to reason that this is always safe in this circumstance, since the regex should protect you. If another developer comes along and changes the regex, but doesn't understand that you've made this assumption, it may fail at runtime.

Not such a big deal here, but reasoning about when a method should be safe is harder than either using a method that explicitly models a missing or bad value (e.g. `head` vs `headOption`) or just wrapping in a `Try().toOption` or similar.

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/Component.scala:
```diff
@@ -24,9 +25,9 @@ final case class Component(idOverride: Option[String],
-                           exposedPorts: List[String] = List())
-    extends ProjectElement
-    with Service {
+                           exposedPorts: List[String] = List(),
```
These should be `PortMapping` instead of `String` now.

NoxHarmonium replied:
Fixed in c2dd744

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/Resource.scala:
```diff
@@ -9,17 +9,23 @@ import com.typesafe.config.Config
+                          settings: Option[Config],
+                          image: String,
+                          environment: Map[String, String] = Map(),
+                          exposedPorts: List[String] = List(),
```
`List` -> `immutable.Seq`, `List()` -> `Nil`.

NoxHarmonium replied:
Is `immutable.Seq` preferable to `Nil` in this case?

dspasojevic commented on src/test/scala/au/com/agiledigital/toolform/model/PortMappingTest.scala:
```diff
@@ -0,0 +1,48 @@
+
+import org.scalatest.{FlatSpec, Matchers}
+
+class PortMappingTest extends FlatSpec with Matchers {
```
These would be good candidates for property based tests: https://www.scalacheck.org

You have a particularly good "there and back again" option, because any `PortMapping`'s string representation should be validly parsed into a `PortMapping` equal to the original.

NoxHarmonium replied:
I added ScalaCheck tests for `String` -> `PortMapping` -> `String` and `PortMapping` -> `String` -> `PortMapping`. I'm sure I could think up some more if I read into the ScalaCheck documentation for other ideas but I think this will be enough for now in the interest of time.
Tests were added in 770cf69

NoxHarmonium commented:
@dspasojevic Thanks for the great feedback!

I've addressed all your comments and it is ready for review again.

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/model/PortMapping.scala:
```diff
@@ -0,0 +1,74 @@
+    * Returns the formatted string representation of this object that can be parsed back in again with the
+    * parsePortMappingFromConfigString function. It is compatible with Docker Compose port format.
+    *
+    * @see https://docs.docker.com/compose/compose-file/#short-syntax-1
```
Nice.

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/app/ToolFormApp.scala:
```diff
@@ -64,7 +63,10 @@ object ToolFormApp extends App {
             .action((_, c) => c.copy(generateTaskConfiguration = c.generateTaskConfiguration.copy(generateTaskOutputType = GenerateTaskOutputType.DockerComposeV3)))
-            .text("generate a Docker Compose v3 file as output (default)")
+            .text("generate a Docker Compose v3 file as output (default)"),
+          opt[Unit]('k', "generate-kubernetes")
```
Will need to be converted to decline sub-commands. Expect invocation to look like:

`toolform generate minikube -i ... -o ...`

`toolform generate --help` should like the available generators. This should work if decline is correctly configured.

@timechanter are you happy with that output?

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/kubernetes/minikube/GenerateMinikube.scala:
```diff
@@ -0,0 +1,60 @@
+      // Final state needs to be read for anything to happen because of lazy evaluation
+      val _ = writeFile.run(context).value
+
+      Right("Completed Successfully")
```
"Wrote configuration to [$outFile], run with ???" ?

NoxHarmonium replied:
Fixed in 4b01e1c

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/tasks/generate/kubernetes/minikube/GenerateMinikube.scala:
```diff
@@ -0,0 +1,60 @@
+
+      Right("Completed Successfully")
+    } finally {
+      writer.close()
```
Note that there are some automatic resource management libs for scala that can make this a little nicer. No need to change for this PR.

NoxHarmonium replied:
Sounds interesting!

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/util/ConfigHelpers.scala:
```diff
@@ -0,0 +1,24 @@
+    * - `Right(t)` becomes `_ => Right(t)`
+    * - `Left(message)` becomes `location => Left(CannotConvert(value, type, message, location)`
+    */
+  def handleErrorWithEither[T](f: String => Either[String, T])(implicit ct: ClassTag[T]): String => Option[ConfigValueLocation] => Either[CannotConvert, T] =
```
`CannotConvert` => `ConfigReaderFailure` ?

NoxHarmonium replied:
Fixed in a77d0f4

dspasojevic commented on src/main/scala/au/com/agiledigital/toolform/util/ConfigHelpers.scala:
```diff
@@ -0,0 +1,24 @@
+  def handleErrorWithEither[T](f: String => Either[String, T])(implicit ct: ClassTag[T]): String => Option[ConfigValueLocation] => Either[CannotConvert, T] =
+    string =>
+      location =>
+        f(string) match {
```
Prefer `leftMap`.

NoxHarmonium replied:
That is much cleaner!

Fixed in a77d0f4

dspasojevic reviewed (changes requested):
Close. Will need to rebase off the decline changes.

ewheath reviewed (commented):
Nothing major to add.  Looks good.  Only suggestion is that we might want to look at reducing duplication in the tests in a separate PR since there are still a few more formats to add...small stuff like pathToFile and common input files.

NoxHarmonium commented:
Do we want every output type of Generate to be a separate plugin or will Generate be a single plugin containing all the output types?

I ask because I currently have them as separate plugins but I can't seem to get them to work under the same command.

For example:

Plugin A:

```
Opts.subcommand("generate", "generates config files for container orchestration.") {
      Opts.subcommand("minikube", "generates config files for Kubernetes (Minikube) container orchestration.") {
     ...
    }
}
```

Plugin B:

```
Opts.subcommand("generate", "generates config files for container orchestration.") {
      Opts.subcommand("dockercompose", "generates config files for container orchestration.") {     ...
    }
}
```

The output of `generate --help` is:

```
Usage: toolform generate minikube

generates config files for container orchestration.

Options and flags:
    --help
        Display this help text.

Subcommands:
    minikube
        generates config files for Kubernetes (Minikube) container orchestration.
```

Only one of them actually get put under the "generate" command. I think that `decline` just uses the last subcommand to register with a given name.

This problem would be solved easily if all the output types were under the same plugin. Otherwise we will have to try a different way to do it. Maybe even examine the source of `decline` to see what happens when you register two commands with the same name.

dspasojevic commented:
@NoxHarmonium `generate` commands should be added by a separate type of plugin.  Just confirm that decline supports nesting sub-commands first. 

Minimal skeleton to test nesting:

```
import java.nio.file.Path

import cats.implicits._
import com.monovore.decline._

/**
  * A tool to generate CI/CD pipelines from a project definition.
  */
object ToolFormApp
    extends CommandApp(
      name = "hello-world",
      header = "Says hello!",
      main = InspectCommand.command orElse GenerateCommand.command
    )

/**
  * A simple error type for the toolform CLI app.
  *
  * @param message The error detail message.
  */
final case class ToolFormError(message: String)

trait ToolformPlugin {
  def command: Opts[Unit]
}

object InspectCommand extends ToolformPlugin {

  val inputFile = Opts.option[Path]("input", short = "i", metavar = "file", help = "Input file")

  val inspectOptions = inputFile.map { inputFile =>
    println(s"Should be inspecting [$inputFile]")
  }

  override val command: Opts[Unit] = Opts.subcommand("inspect", "Inspect and print the content of the project file")(inspectOptions)
}

object GenerateCommand extends ToolformPlugin {
  val quietOpt = Opts.flag("quiet", help = "Whether to be quiet.").orFalse

  val inputFile = Opts.option[Path]("input", short = "i", metavar = "file", help = "Input file")

  override val command: Opts[Unit] = Opts.subcommand(
    "generate",
    "Generate the tool configuration for the project file"
  ) {
    (inputFile |@| quietOpt).map { (inputFile, quiet) =>
      println(s"Should be generating from [$inputFile] with q [$quiet]")
    }
  }
}
```

NoxHarmonium commented:
@dspasojevic Nested commands is supported fine. It just doesn't support declaring the parent command in each plugin.

So we would need code that declares the generate subcommand, and then imports all the generate sub commands in that scope.

dspasojevic commented:
@NoxHarmonium yes, exactly. A generate plugin (peer to inspect) that adds the generate command. Then separate plugin type that the generate sub-commands implements.

NoxHarmonium commented:
Ok I will try that

Commit:
[14] Implement generate subtask handling

Commit:
[14] Add better output information

Commit:
[14] Use left.map to clean up code

Commit:
[14] Fix tests broken in rebase

NoxHarmonium commented:
@dspasojevic I've rebased and addressed your feedback.

The rebase was messy so I squashed it down to one commit. I hope it doesn't make reviewing it a lot harder.

Commit:
[14] Update readme and homogenise help text

dspasojevic reviewed (approved):
Nice.
